src/MessageHandler.cpp

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2005 - 2015 Settlers Freaks (sf-team at siedler25.org)
+//
+// This file is part of Return To The Roots.
+//
+// Return To The Roots is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// Return To The Roots is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Return To The Roots. If not, see <http://www.gnu.org/licenses/>.
+
+#include "libUtilDefines.h"
+#include "MessageHandler.h"
+#include "Message.h"
+#include "Socket.h"
+#include "SocketSet.h"
+#include "Log.h"
+#include "MyTime.h"
+
+int MessageHandler::send(Socket& sock, const Message& msg)
+{
+    Serializer ser;
+    ser.PushUnsignedShort(0); // Id
+    ser.PushSignedInt(0); // Placeholder for length
+    const unsigned headerSize = ser.GetLength();
+    msg.Serialize(ser);
+
+    if(ser.GetLength() > 64 * 1024) /// 64KB, ACHTUNG: IPV4 garantiert nur maximal 576!!
+    {
+        LOG.lprintf("BIG OOPS! Message with length %u exceeds maximum of %d!\n", ser.GetLength(), 64 * 1024);
+        return -1;
+    }
+
+    unsigned char* data = ser.GetDataWritable(ser.GetLength());
+    *reinterpret_cast<int*>(data) = msg.getId();
+    *reinterpret_cast<int*>(data + sizeof(unsigned short)) = ser.GetLength() - headerSize;
```

Comment:
The problem is that it would not be streaming. PushFront requires moving all succeeding data backwards which would be some serious overhead.
Note that the id COULD be pushed above, but Push cannot be used due to https://github.com/Return-To-The-Roots/s25client/issues/363